### PR TITLE
[google_maps_flutter] Implement google maps flutter heatmap

### PR DESCRIPTION
**Pull request by SvenSlijkoord**
## Description

Implements heatmap support for the google_maps_flutter package.

## Related Issues

flutter/flutter#33586

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by urmilshroff** [comment]
Thank you for this! Let's hope it gets merged soon 🙏

File: packages/google_maps_flutter/example/ios/Flutter/Flutter.podspec [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  s.description      = <<-DESC
+Flutter provides an easy and productive way to build and deploy high-performance mobile apps for Android and iOS.
+                       DESC
+  s.homepage         = 'https://flutter.io'
+  s.license          = { :type => 'MIT' }
+  s.author           = { 'Flutter Dev Team' => 'flutter-dev@googlegroups.com' }
+  s.source           = { :git => 'https://github.com/flutter/engine', :tag => s.version.to_s }
+  s.ios.deployment_target = '8.0'
+  s.vendored_frameworks = 'Flutter.framework'
+end
```
Comment (cyanglaz) [inline packages/google_maps_flutter/example/ios/Flutter/Flutter.podspec]: revert

File: packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+static double ToDouble(NSNumber* data) { return [FLTGoogleMapJsonConversions toDouble:data]; }
+
+static BOOL ToBool(NSNumber* data) { return [FLTGoogleMapJsonConversions toBool:data]; }
+
+static NSArray<GMUWeightedLatLng*>* ToPoints(NSArray* data) {
+  NSMutableArray* points = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    NSNumber* latitude = data[i][0][0];
+    NSNumber* longitude = data[i][0][1];
+    NSNumber* intensity = data[i][1];
```
Comment (cyanglaz) [inline packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m]: will any of these lines crash? For example, will `data[i].length < 2` and thus crashes `data[i][1]`

File: packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+static NSArray<UIColor*>* ToColors(NSArray* data) {
+  NSMutableArray* colors = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    UIColor* color = data[i];
+    [colors addObject:color];
+  }
+
+  return colors;
```
Comment (cyanglaz) [inline packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m]: The method seems unnecessary. It doesn't seem to be doing anything with the data in the array. The method just did a deep copy from `data` to the return value.

File: packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+static NSArray<NSNumber*>* ToStartsPoints(NSArray* data) {
+  NSMutableArray* startPoints = [[NSMutableArray alloc] init];
+  for (unsigned i = 0; i < [data count]; i++) {
+    NSNumber* startPoint = data[i];
+    [startPoints addObject:startPoint];
+  }
+
+  return startPoints;
+}
```
Comment (cyanglaz) [inline packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m]: The method seems unnecessary. The method seems unnecessary. It doesn't seem to be doing anything with the data in the array. The method just did a deep copy from `data` to the return value.

**Comment by cyanglaz** [comment]
I think it might be a good idea to separate tileoverlay and heatmap into 2 different PRs. This will help the PRs to be reviewed faster. https://github.com/flutter/plugins/pull/2077 seems trying to add tile overlay for the plugin on Android. @SvenSlijkoord Do you think you can work with the author to put up a PR for tile overlay only? And then a separate PR for heat map?

**Comment by SvenSlijkoord** [comment]
@cyanglaz let me look into it.

File: packages/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 5]
```diff
@@ -197,6 +199,15 @@ static Object circleIdToJson(String circleId) {
     return data;
   }
 
+  static Object heatmapIdToJson(String heatmapId) {
```
[the remark this answers is not in the document]
Comment (bmabir17) [reply packages/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:202]: yes, its not used. It seems he tried to follow same pattern as circleController.

**Comment by Alystrasz** [comment]
Are there any updates on this PR? Are CI-reported issues hard to fix?
Heatmaps feature is part of the [flutter todo list](https://github.com/flutter/flutter/projects/110#card-32220384) and is long-awaited by devs (https://github.com/flutter/flutter/issues/33586).

**Comment by bmabir17** [comment]
> Are there any updates on this PR? Are CI-reported issues hard to fix?
> Heatmaps feature is part of the [flutter todo list](https://github.com/flutter/flutter/projects/110#card-32220384) and is long-awaited by devs ([flutter/flutter#33586](https://github.com/flutter/flutter/issues/33586)).
@Alystrasz 
I urgently needed this feature so i have forked this pull request
And published a forked package [google_maps_flutter_heatmap](https://pub.dev/packages/google_maps_flutter_heatmap) (Tested only on android)

**Comment by Alystrasz** [comment]
@bmabir17 
Actually, I'm using your fork right now in my project (thanks btw), but I need iOS support as well :smiley:

**Comment by bmabir17** [comment]
> @bmabir17 
> Actually, I'm using your fork right now in my project (thanks btw), but I need iOS support as well :smiley: 

Can tell me what's the error that you are facing for ios? Don't have an ios device to debug it 🙁

File: packages/google_maps_flutter/lib/src/controller.dart [hunk 6]
```diff
@@ -164,6 +164,20 @@ class GoogleMapController {
...
+  /// Change listeners are notified once the update has been made on the
+  /// platform side.
+  ///
+  /// The returned [Future] completes after listeners have been notified.
+  Future<void> _updateHeatmaps(_HeatmapUpdates heatmapUpdates) async {
+    assert(heatmapUpdates != null);
+    await channel.invokeMethod<void>(
+      'heatmaps#update',
+      heatmapUpdates._toMap(),
+    );
```
Comment (ditman) [inline packages/google_maps_flutter/lib/src/controller.dart:178]: This should become a method in the `platform_interface` of the plugin.

File: packages/google_maps_flutter/lib/src/heatmap.dart [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of google_maps_flutter;
+
+/// Uniquely identifies a [Heatmap] among [GoogleMap] heatmaps.
+///
+/// This does not have to be globally unique, only unique among the list.
+@immutable
+class HeatmapId {
```
Comment (ditman) [inline packages/google_maps_flutter/lib/src/heatmap.dart:11]: These classes should be moved to the `platform_interface` package

File: packages/google_maps_flutter/lib/src/heatmap_updates.dart [hunk 8]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of google_maps_flutter;
+
+/// [Heatmap] update events to be applied to the [GoogleMap].
+///
+/// Used in [GoogleMapController] when the map is updated.
+class _HeatmapUpdates {
```
Comment (ditman) [inline packages/google_maps_flutter/lib/src/heatmap_updates.dart:10]: These classes should live in the `platform_interface` package.

File: packages/google_maps_flutter/test/fake_maps_controllers.dart [hunk 9]
```diff
@@ -96,6 +103,9 @@ class FakePlatformGoogleMap {
       case 'circles#update':
         updateCircles(call.arguments);
         return Future<void>.sync(() {});
+      case 'heatmaps#update':
+        updateHeatmaps(call.arguments);
+        return Future<void>.sync(() {});
```
Comment (ditman) [inline packages/google_maps_flutter/test/fake_maps_controllers.dart:108]: This should be converted to a `Stream<HeatmapUpdateEvent>` in the `platform_interface` (all the code of onMethodCall has migrated to that package as well).

**Comment by YazeedAlKhalaf** [comment]
Guys, When will this be merged?

**Comment by ditman** [comment]
The refactor has [landed](https://github.com/flutter/plugins/commit/1a9d51d344f71b0f8af657053b04783e274fd1ca).

**Comment by Alystrasz** [comment]
@SvenSlijkoord do you plan to refactor your PR according to the recent plugins refactor ?

**Comment by SvenSlijkoord** [comment]
Yes, I will

File: packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m (hunk 4 above)
Comment (SvenSlijkoord) [reply packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m]: This method converts a JSON object (NSArray* data) to a array of NSNumber*. I am no iOS developer but it didn't seem to work without this method.

File: packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m (hunk 3 above)
Comment (SvenSlijkoord) [reply packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m]: This method converts a JSON object (NSArray* data) to an array of UIColor*. I am no iOS developer but it didn't seem to work without this method.

File: packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m (hunk 2 above)
Comment (SvenSlijkoord) [reply packages/google_maps_flutter/ios/Classes/GoogleMapHeatmapController.m]: That would most definitely crash, but that would also mean that the flutter part of this package is implemented improperly.
Should we throw an exception when the data is not correct, or how can this ideally be resolved?

File: packages/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SvenSlijkoord) [reply packages/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java]: This was indeed not used and is now removed.

**Comment by stuartmorgan** [comment]
Once the migration of maps to null-safety is complete, we can move forward with https://github.com/flutter/plugins/pull/3062, which takes on just the interface portion of this task, and then hopefully this can be updated to have a narrower scope, and we can get it reviewed.

**Comment by stuartmorgan** [comment]
We're starting the process of reviewing #3062. While that's happening, this could be updated in parallel to add tests (integration tests, unit tests for the native code) so that it will be closer to being review-ready once that PR is complete and landed.

**Comment by stuartmorgan** [comment]
Status update from PR triage. https://github.com/flutter/plugins/pull/3062 is still going through review.

However, this still needs tests before it could move forward even once that PR has landed.

**Comment by ditman** [comment]
> Status update from PR triage: https://github.com/flutter/plugins/pull/3062 is still going through review.

@stuartmorgan The platform_interface on #3062 has diverged so much from the one here, that IMO we need to bring this PR up to date with the latest code there, and see how it works.

**Comment by stuartmorgan** [comment]
That makes sense. @SvenSlijkoord Are you still interested in updating this PR to the current state of the plugin?

**Comment by stuartmorgan** [comment]
Thank you for your contribution. I'm going to close this PR for now since there are outstanding comments, just to get this off our PR review queue. Please don't hesitate to submit a new PR if you have the time to address the review comments. Thanks!

**Pull request closed without merging**